File: python_solutions/day_6.py
from pathlib import Path


INPUT_FILE_PATH = Path("../inputs/day_6.txt")


def get_signal() -> str:
    with open(INPUT_FILE_PATH, "r") as f:
        return f.read()


def search_for_marker(signal: str, marker_size: int) -> int:
    for i in range(len(signal) - (len(signal) % marker_size) - marker_size):
        if len(set(signal[i : i + marker_size])) == marker_size:
            return i + (marker_size - 1)
    raise RuntimeError("No marker found")


def main():

    signal = get_signal()
    part_1 = search_for_marker(signal, marker_size=4)
    part_2 = search_for_marker(signal, marker_size=14)

    print(f"Packet received at character {part_1 + 1}")
    print(f"Message received at character {part_2 + 1}")


if __name__ == "__main__":
    main()

File: cpp_solutions/day_6.cpp
#include <string>
#include <fstream>
#include <set>
#include <exception>
#include <iostream>

#define INPUT_FILE_PATH "../inputs/day_6.txt"

std::string getSignal() {
    std::ifstream file(INPUT_FILE_PATH);
    std::string line;
    std::getline(file, line);
    return line;
}

int searchForMarker(std::string signal, int markerSize) {
    int remainder = ((signal.length() % markerSize) + markerSize) % markerSize;
    for (int i = 0; i < (signal.length() - remainder - markerSize); i++) {
        std::set<char> chunk;
        for (int j = 0; j < markerSize; j++) {
            chunk.insert(signal.at(i + j));
        }
        if (chunk.size() == markerSize) {
            return i + (markerSize - 1);
        }
    }
    throw std::runtime_error("No marker found");
}

int main() {
    std::string signal = getSignal();
    int part1 = searchForMarker(signal, 4);
    int part2 = searchForMarker(signal, 14);
    std::cout << "Packet received at character " << part1 + 1 << std::endl;
    std::cout << "Message received at character " << part2  + 1 << std::endl;
}